letsencrypt/models/letsencrypt.py

```diff
@@ -0,0 +1,157 @@
+        domain_key = os.path.join(self.get_data_dir(), '%s.key' % domain)
+        if not os.path.isfile(domain_key):
+            _logger.info('generating rsa domain key for %s', domain)
+            self.call_cmdline([
+                'openssl', 'genrsa', '-out', domain_key,
+                str(DEFAULT_KEY_LENGTH),
+            ])
+        return domain_key
+
+    @api.model
+    def validate_domain(self, domain):
+        if domain in ['localhost', '127.0.0.1'] or\
+                domain.startswith('10.') or\
+                domain.startswith('172.16.') or\
+                domain.startswith('192.168.'):
+            raise exceptions.Warning(
+                _("Let's encrypt doesn't work with private addresses!"))
+
+    @api.model
+    def generate_csr(self, domain):
+        domains = [domain]
+        i = 0
+        while self.env['ir.config_parameter'].get_param(
+                'letsencrypt.altname.%d' % i):
+            domains.append(
+                self.env['ir.config_parameter']
+                .get_param('letsencrypt.altname.%d' % i)
+            )
+            i += 1
+        _logger.info('generating csr for %s', domain)
+        if len(domains) > 1:
+            _logger.info('with alternative subjects %s', ','.join(domains[1:]))
+        config = self.env['ir.config_parameter'].get_param(
+            'letsencrypt.openssl.cnf', '/etc/ssl/openssl.cnf')
+        csr = os.path.join(self.get_data_dir(), '%s.csr' % domain)
+        with tempfile.NamedTemporaryFile() as cfg:
+            cfg.write(open(config).read())
+            if len(domains) > 1:
+                cfg.write(
+                    '\n[SAN]\nsubjectAltName=' +
+                    ','.join(map(lambda x: 'DNS:%s' % x, domains)) + '\n')
+            cfg.file.flush()
+            cmdline = [
+                'openssl', 'req', '-new',
+                self.env['ir.config_parameter'].get_param(
+                    'letsencrypt.openssl.digest', '-sha256'),
+                '-key', self.generate_domain_key(domain),
+                '-subj', '/CN=%s' % domain, '-config', cfg.name,
+                '-out', csr,
+            ]
+            if len(domains) > 1:
+                cmdline.extend([
+                    '-reqexts', 'SAN',
+                ])
+            self.call_cmdline(cmdline)
+        return csr
+
+    @api.model
+    def cron(self):
+        domain = urlparse.urlparse(
+            self.env['ir.config_parameter'].get_param(
+                'web.base.url', 'localhost')).netloc
+        self.validate_domain(domain)
+        account_key = self.generate_account_key()
+        csr = self.generate_csr(domain)
+        manifest, manifest_path = file_open(
+            'letsencrypt/__openerp__.py', pathinfo=True)
+        manifest.close()
+        acme_challenge = os.path.join(
+            os.path.dirname(manifest_path),
+            'static', 'acme-challenge')
+        if not os.path.isdir(acme_challenge):
+            os.makedirs(acme_challenge)
```

Comment:
This chunk of code is quite weird.

Are you opening the `__openerp__.py` file just to get the module's path to create a folder inside it? :hushed:

First, you have the `__file__` magical variable to know your addon's dir, and second you should never write there. Addons should ideally be mounted in readonly folders. You have the filestore for writing, in a per-database basis.